src/build.rs

```diff
@@ -47,27 +52,76 @@ fn ensure_nightly() -> Result<(), Error> {
 
 /// Run `cargo build` with the `nightly` toolchain and targetting
 /// `wasm32-unknown-unknown`.
-pub fn cargo_build_wasm(path: &Path, debug: bool, step: &Step) -> Result<(), Error> {
+///
+/// Returns the location of the built wasm file.
+pub fn cargo_build_wasm(
+    path: &Path,
+    debug: bool,
+    step: &Step,
+    crate_name: &str,
+) -> Result<PathBuf, Error> {
     let msg = format!("{}Compiling to WASM...", emoji::CYCLONE);
     PBAR.step(step, &msg);
-    let output = {
-        let mut cmd = Command::new("cargo");
-        cmd.current_dir(path)
-            .arg("+nightly")
-            .arg("build")
-            .arg("--lib");
-        if !debug {
-            cmd.arg("--release");
-        }
-        cmd.arg("--target").arg("wasm32-unknown-unknown");
-        cmd.output()?
-    };
 
-    if !output.status.success() {
-        let s = String::from_utf8_lossy(&output.stderr);
-        Error::cli("Compilation of your program failed", s)
+    let mut cmd = Command::new("cargo");
+    cmd.current_dir(path)
+        .arg("+nightly")
+        .arg("build")
+        .arg("--lib")
+        .arg("--target=wasm32-unknown-unknown")
+        .arg("--message-format=json")
+        .stderr(Stdio::piped())
```

Comment:
For this I believe it will need to pessimistically spawn a new thread to read stderr. Othrewise the blocking reads on stdout could cause wasm-pack to deadlock when it's waiting for info on stdout but the child is waiting to write info to stderr.

Reply:
Afaik we're spawning a new process here and stderr is new pipe from the os, so the child can write to stderr no matter what the parent process does (except broken pipe errors and the like). Not an expert though.

Reply:
Pipes have a fixed capacity, however, and won't infinitely buffer. If Cargo dumps a lot of information on stderr it'll eventually block, and then it won't make progress until the pipe is read, which wasm-pack doesn't do until the process has otherwise finished.

Reply:
Ah thanks, I didn't think about that! I've replaced the pipe with a tempfile.